### DISCUSSION TITLE
Mempool Refactor

### DISCUSSION THREAD

**Discussion by alexanderbez** [tendermint/tendermint#6295] [General]
There have been many discussions about the current mempool design, its limitations and various design proposals for improving it. Having a central place for discussion should prove to be helpful until all relevant stakeholders and parties have agreed upon a design and overall approach.

To quickly recap, the current mempool implementation in Tendermint is a simple thread-safe FIFO queue -- the order in which txs come into a proposer's mempool is the order in which they make it into a block and go out, respecting gas limits of course.

It should be pretty clear to see that this approach is very limiting for a number of reasons. Specifically, there is zero prioritization, e.g. a common use case would be for ABCI-enabled applications to prioritize and order txs by fees, preferring higher fee txs. The lack of prioritization also leads to cascading effects in terms of DoS and various attack vectors on networks, e.g. https://github.com/cosmos/cosmos-sdk/discussions/8224.

The two primary proposals for a revised mempool include:

1. Modified `CheckTx`
    - In this proposal, the idea is to keep things as simple as possible, requiring the least amount of change, at least at the ABCI boundary. The proposal entails having `CheckTx` return a `priority` value and Tendermint would use a data structure to store txs ordered by priority, e.g. a simple thread-safe priority queue. Note, there would be a requirement for applications to also keep track of pending txs due to the fact that we need to have two-dimensional ordering as to ensure txs are processed in the correct order with respect to account nonces, e.g.` <nonce, fees>`.
2. ABCI++
    - In this proposal, the changes are more complex and come with various changes to the ABCI interface and protocol. Specifically, we leverage ABCI++'s `PrepareProposal` ABCI method. In this method, the application tells Tendermint what txs to include in the next block. This gives greater flexibility to the application compared to (1). Within Tendermint, the mempool would become a simple map, mapping from a unique tx identifier, e.g. hash, to the tx itself.
    - It should be noted, that once the ABCI++ spec is completed and consensus is reached, the mempool work can begin in piece-meal fashion as the other ABCI changes part of ABCI++ can be no-ops.

cc @marbar3778 @ValarDragon @sunnya97 @ebuchman

**Comment by tac0turtle** [comment]
An added caveat to the mempool that ABCI++ enables, we will need to solve with the refactor that has gone a bit under the radar.

When a proposer is proposing a block he can modify the transaction, thus changing the hashes. When we go to remove txs from the mempool  the hashes of the txs would not match, thus causing us to checkTx every tx in order to find which tx was included in the block. We could introduce a cache of the txs added to the block and if the block is finalized we use that hash to eject the txs from the mempool.

The modifying of transaction is a gotcha to us, but it allows major performance benefits app side.

**Reply by alexanderbez** [reply]
Good point, but the mempool in this case just becomes a simple thread-safe cache, so is an extra cache necessary? Also, do we have to use tx hashes? Is using a UUID or something similar possible?

**Reply by ValarDragon** [reply]
I want to point out that this fast scan for removing already seen txs is very much a second order optimization imo, that doesn't actually have an impact at large mempool sizes. (And can be eliminated with specifying `sender, seqnum` in the CheckTx response)

The decision criterion imo should be
1) Don't care about this optimization for now, its of never-benchmarked benefit. (Mattered more prior to recheck being implemented)
2) Make the app return `sender, seqnum`, and have the mempool have invalidation logic. (If a block contains a tx that passed execution with `sender_A, seqnum_A` and mempool contains tx from `sender_A, seqnum_B`, if `seqnum_A >= seqnum_B` invalidate the tx)
3) Make a "witness-independent-tx hash", basically segwit for txhash's. You need this anyway for checking if you're tx hash was included within a block in settings where it gets mutated, this would just be making the mempool aware of it.

**Reply by ebuchman** [reply]
What's the use case for having the ABCI app modify txs? Is it like signature aggregation? Anything else?

PrepareProposal could also just return the hashes of the txs it included if it modified them

**Reply by ebuchman** [reply]
also are ResponseCheckTx changes and ABCI++ really exclusive? I would think we actually want both, and PrepareProposal could over ride what ever the default Tendermint does ? Of course we'd probably start with the the ResponseCheckTx anyways, but maybe want to ensure it will work well with PrepareProposal when it's ready

**Comment by tac0turtle** [comment]
After discussion with @ValarDragon in slack, the first option seems to be the best to move forward with. 

We could start working on the refactor when @alexanderbez is back from vacation.

**Reply by ValarDragon** [reply]
In particular, there was suggestion of an approach where only the application maintained the priority, and prepareProposal of ABCI++ would handle selecting the highest priority things and placing them into a block.

However this would not solve the problem of tx gossip under load. Suppose there were tons of txs in the mempool, and I had a tx with a fee of say 10 atoms. Then we want my tx to be prioritized in gossip so it gets to a validator ASAP. Consequently the code that handles tx gossiping must be aware of the priority. So if only the application handles knowing the priority, it must handle selecting txs to gossip. This is not the design goal at the moment. (Thats a fully app-defined mempool) 

Thus the TM mempool must know the tx priorities, in order to gossip high priority txs first.

**Comment by zmanian** [comment]
Do we think replace by fee is an important property to have in the memory?

Do ABCI semantics need to be able to say this tx replaces another tx in the queue?

**Reply by tac0turtle** [reply]
I think this is a cool feature. Some ways it could be done is via recheck (not turned on by all nodes) or introducing an inverted ABCI interface which allows the app to send requests to Tendermint, in this case a replace tx message.

**Reply by ValarDragon** [reply]
Replace by fee is possible if the state machine outputs `(sender, seqnum, priority)` in the ABCI checkTx response. Also useful as a 'fast duplicate elimination' check (2. in https://github.com/tendermint/tendermint/discussions/6295#discussioncomment-545266). This also suffices for UTXO models, albeit with some additional data overhead over ABCI.

This would require changes to the underlying data structures though. (You may have to double the 'data structural' overhead, so that you can sort by priority, and search by `(sender, seqnum)`. Basically what you do in SQL databases to have multiple columns, where each leaf of the sorted tree then points to the tx object to remove duplication) 

I propose that we make `(sender, seqnum, priority)` the ABCI response, and first build a mempool sorted by fees. Then we add replace by fee semantics later, as the interfaces would already suffice.

also, recheck not being done for all nodes is mainly an artifact of certain optimizations not existing before, which we can likely remove with some more mempool optimizations. (In particular, when there was no recheck ABCI mode for CheckTx when it was added, and https://github.com/tendermint/tendermint/issues/6232 would fix the remaining concerns)

**Reply by alexanderbez** [reply]
The data structures will change anyway, so I don't think that's a concern.

Are we all in agreement that we're OK with adding `(sender, nonce, priority)` to the `CheckTx` response?

**Reply by zmanian** [reply]
I like it

**Reply by ValarDragon** [reply]
I like it as well. I just want to ensure that the Tendermint side is opaque w.r.t. what the value of `sender` is, and it only cares about its uniqueness. (e.g. using account number in Account models, and address in UTXO models)

**Reply by alexanderbez** [reply]
Yeah it can just be a string of whatever value or even just bytes? wdyt?

**Reply by ebuchman** [reply]
What does `sender` mean in a UTXO context if we're talking about a whole transaction (ie. with multiple inputs) ? Also whats the nonce for a UTXO ? Empty? 

I think we'd ultimately want both replace-by-fee and child-pays-for-parent. But I think in both cases we need some work on the SDK side to support rolling back part of the state (or something like keeping a CacheKV store for each sender :o ).

**Reply by alexanderbez** [reply]
For UTXOs there isn't a native sender, so this wouldn't work I think? Nonces can be omitted.

What do folks think about UTXO-based applications? Is there an approach that covers both types?

**Reply by ValarDragon** [reply]
I think there is a native sender, its moreso that its a list of senders in the case of a multi-input tx

**Reply by alexanderbez** [reply]
@ValarDragon @ebuchman do you have suggestions on how to approach this? As far as I see it:

1. We either have a mechanism where Tendermint can find out what is the next processable tx, OR
2. Only allow 1 tx from 1 sender at a time, i.e. no clist

Am I missing other alternatives?

**Comment by JayT106** [comment]
Question for the mempool design, 
1. Any maximum allowance staying time settings to let the transaction drop from the mempool? i.e. automatically drop transactions if transactions stuck in the pool too long. 
2. Any special cases the transaction has been added into the mempool (current implementation) but somehow been dropped later? our exchange team has concerns for this case.

**Reply by alexanderbez** [reply]
Great questions @JayT106! We had not considered these cases or concerns. To address them,

Point (1), we had not considered this and my feeling is that this is really out of scope of the changes we're proposing but certainly something to keep in mind. My initial idea on how to approach this would actually have nothing to do with the proposal at all, all we would need to do is have a new mempool configuration, e.g. `max_mempool_duration`, and we could kick out txs that stay longer than this. Can you create an issue for this?  It's just out of scope of the work being discussed here.

I dont quite understand what you mean by point (2), can you elaborate a bit please?

**Reply by JayT106** [reply]
@alexanderbez thanks for the feedback,
For point 1, agree, this is out of scope in this proposal, I will create an issue to address the issue.
For point 2, our team just wants to know any chance the transaction could be dropped in the mempool somehow(it was an valid tx in the mempool). I looked at the mempool implementation it looks fine but you might know the mempool has some legacy issue related to this concern.

**Reply by ValarDragon** [reply]
In the current implementation, a tx that has succesfully made it into the mempool will only get dropped if the application says its no longer valid. (This would primarily happen due to sequence numbers incrementing)

**Comment by alexanderbez** [comment]
@zmanian @ValarDragon what do you think of https://github.com/tendermint/tendermint/blob/bez/adr-067-mempool-refactor/docs/architecture/adr-067-mempool-refactor.md#detailed-design so far?

**Reply by ebuchman** [reply]
Thanks for getting this going Bez great start left some comments in this discussion.

**Reply by alexanderbez** [reply]
Thanks for reviewing @ebuchman!

**Comment by ebuchman** [comment]
From the [draft](https://github.com/tendermint/tendermint/blob/bez/adr-067-mempool-refactor/docs/architecture/adr-067-mempool-refactor.md#detailed-design):

> The existing concurrent-safe linked-list will be removed entirely in favor of a thread-safe map of <sender:[]Tx>, i.e a mapping from sender to a list of Tx objects

Note we may still want to use concurrent-safe linked lists here to manage the peer gossip. Right now the clist ensures that all txs are sent to each peer in order, and the list of txs can be dynamically modified (ie. txs removed from anywhere and txs pushed onto the top). 

It's possible the need for sending in order is resolved by buffering txs that can't be executed yet, in which case the clist would just be an optimization (I think, but we should probably also mention this in the ADR). As described in the ADR:

> For example, the sender Alice, whose current committed sequence value is 3, could have transactions in this map with sequence values <4,5,6,8>, where only the transaction with sequence 4 could be processed next. 

But then it also says:

> Note, the mempool itself does not and will not know about the latest committed sequence value, but it is assumed that all transactions that end up in the mempool pass CheckTx

Does this mean a tx would be expected to pass CheckTx even if the nonce was too high? But then couldn't we end up putting invalid txs in blocks? Eg. if the nonce for a sender was 3 and we receive txs with nonce 5,6,7 and they pass CheckTx and we put them in a block ? So I think if we want to allow the nonce buffering we need some way to figure out the sequence number for the account, OR, possibly, check the error code on CheckTx and if its InvalidSequence or whatever than just buffer the tx for later.

We probably also want to include something in the ADR about a limit on the number of `sender` buckets we allow and/or their size (both total size and number of buffered txs that cant be executed yet).

**Reply by alexanderbez** [reply]
> Does this mean a tx would be expected to pass CheckTx even if the nonce was too high? But then couldn't we end up putting invalid txs in blocks?

Yes, this describes a hypothetical application that allows this, i.e imagine the app only cares that they're increasing. So it was meant to be a more relaxed invariant. In the context of the SDK, AFAIK these txs are still valid if you manage the sequence numbers yourself. If you don't, then each `sender` will just have 1 tx in its respective list.

However, you're right that Tendermint might still need to know and this did cross my mind, e.g. if Tendermint gets `<5,6,7>` before it gets `4`. I'm not sure how to handle this case atm :-/

We could remove `sequence` entirely and just allow 1 tx from a sender at a time. Which would simplify things. Thoughts?

**Reply by ebuchman** [reply]
this might be fine for a first pass but i think ultimately we want to support multiple txs from a single sender in the mempool.

I guess the question is if we want to support txs that cant be processed yet (future nonce). I would say its something to keep in mind but probably less important? I think `geth` does this though we could look at their mempool for how its handled but it would prob add a bunch more complexity.

**Reply by alexanderbez** [reply]
Well it's very trivial to support if Tendermint could know the current committed nonce, which is how Ethereum probably does it. Since Tendermint doesn't know, it can't know the true "first processable tx".
